### PR TITLE
Ensure metadata location when created from xml element with known location

### LINKED ISSUE #8579: Metada Location should have nonzero column and line
### Issue Description

`MetadataElement` of an `ItemElement` doesn't have line&col of `Location` set despite those are se for the parent `ItemElement`.

image
[[code](https://github.com/dotnet/msbuild/blob/main/src/Build/Evaluation/Evaluator.cs#L487)]

Errors referencing metadata can hence have location reported as (0,0) <filename>

### Analysis

Location is hardcoded to 0, 0 ([here](https://github.com/dotnet/msbuild/blob/main/src/Build/ElementLocation/XmlElementWithLocation.cs#L33)), because at the time of creating of metadata, the reader is already not available (nullified [here](https://github.com/dotnet/msbuild/blob/main/src/Build/ElementLocation/XmlDocumentWithLocation.cs#L162-L165)), it's only available during Load phase, metadata are created during parsing phase (the two phases [here](https://github.com/dotnet/msbuild/blob/main/src/Build/Construction/ProjectRootElement.cs#L226-L228))

Possible way out: either reuse the Location info from owning parent (possibly off for element definition spanning over multiple lines - but likely still better than (0, 0) location); or rework the code to create metadata during Load phase - or somehow preserve location information from that phase.

### PR DESCRIPTION
**Pull request by JanKrivanek**
Fixes #8579

### Context
`MetadataElement` of an `ItemElement` doesn't have line&col of `Location` set despite those are created from known xml.
This way errors referencing the `Location` from metadata would end up pointing to location (0, 0) in the file.

### Reason
This happens only to `XmlElementWithLocation` (as opposed to `XmlAttributeWithLocation`) during creating `XmlElementWithLocation` from scratch (during parsing phase that doesn't have the reader with locations already available) from an existing `XmlAttributeWithLocation` (that has the proper location intitialized from Load phase). Due to the need to keep the constructors contract (we override existing `XmlElement`), we cannot pass the additional info in the constructor.

### Changes Made
Making the `Location` property of `XmlElementWithLocation` writable - so that it can be properly rewritten if constructed from an existing node with available location info.

### Testing
Hand testing plus applying the changfe to the other PR: https://github.com/dotnet/msbuild/pull/8581, solved the issue with (0, 0) warnings location

### Note
This or https://github.com/dotnet/msbuild/pull/8581 PR (whichever is merged later) should remove the workaround https://github.com/dotnet/msbuild/pull/8581/files#diff-e289ba4ce7fa0e72cf63049cce595eafcad1e7b2034ccb3305cd0f06c2f822b8R196-R197

### CODE REVIEW

**Review by Forgind — APPROVED** [review]
This looks solid to me 🙂

My only real concern is over incorrect locations. I'm mostly thinking of the XmlUtilities change—if its children have locations, then I imagine their locations would shift slightly (longer name --> higher column number; possible reformatting --> different line number; etc.). If those stack up, we might be noticeably wrong, which could be very confusing. Do we appropriately account for those sorts of changes?

**Comment by JanKrivanek** [comment]
@Forgind - very good point about precautions with locations shifts!
The location is set in two places:
* `RenameXmlElement` - there location is preserved from the element being renamed (not the attributes) - location of element should not change with rename
* `ProjectMetadataElement.CreateDisconnected` - there the location is argument is populated only for code paths executed on initial project parsing (`ProjectParser.ParseProjectItemElement`) - so the location is not altered.

As for subsequent edits of the document (and hence possible subsequent locations shift) - this remains unchanged - the location is not guaranteed after edits:

https://github.com/dotnet/msbuild/blob/main/src/Build/ElementLocation/XmlElementWithLocation.cs#L83

**Comment by Forgind** [comment]
> @Forgind - very good point about precautions with locations shifts! The location is set in two places:
> 
> * `RenameXmlElement` - there location is preserved from the element being renamed (not the attributes) - location of element should not change with rename
> * `ProjectMetadataElement.CreateDisconnected` - there the location is argument is populated only for code paths executed on initial project parsing (`ProjectParser.ParseProjectItemElement`) - so the location is not altered.
> 
> As for subsequent edits of the document (and hence possible subsequent locations shift) - this remains unchanged - the location is not guaranteed after edits:
> 
> https://github.com/dotnet/msbuild/blob/main/src/Build/ElementLocation/XmlElementWithLocation.cs#L83

For RenameXmlElement, I think row should be constant, but column could change? Am I wrong about that? Like if I have:
`<Foo><Bar /></Foo>`, then change it to `<FooFooFoo><Bar /></FooFooFoo>`, the column of Bar would change.

**Comment by JanKrivanek** [comment]
> For RenameXmlElement, I think row should be constant, but column could change? Am I wrong about that? Like if I have:
> `<Foo><Bar /></Foo>`, then change it to `<FooFooFoo><Bar /></FooFooFoo>`, the column of Bar would change.

Correct.
This is something that's unchanged by this PR (it's regard the location of other xml elements after editing some elements)

**Comment by Forgind** [comment]
If it's unchanged, I certainly can't complain too much 🙂

I was thinking you were adding locations in places there previously weren't locations, in which case if they're edited --> wrong, users might expect them to be right anyway, whereas they'd have no expectations without a location. But it sounds like you've looked into this fairly carefully, so I'm satisfied with it.

File: src/Build/Construction/ProjectMetadataElement.cs [hunk 1]
```diff
@@ -100,13 +100,17 @@ public string Value
...
+        internal static ProjectMetadataElement CreateDisconnected(string name, ProjectRootElement containingProject, ElementLocation location = null)
         {
             XmlUtilities.VerifyThrowArgumentValidElementName(name);
             ErrorUtilities.VerifyThrowArgument(!FileUtilities.ItemSpecModifiers.IsItemSpecModifier(name), "ItemSpecModifierCannotBeCustomMetadata", name);
             ErrorUtilities.VerifyThrowInvalidOperation(!XMakeElements.ReservedItemNames.Contains(name), "CannotModifyReservedItemMetadata", name);
 
             XmlElementWithLocation element = containingProject.CreateElement(name);
+            if (location != null)
+            {
+                element.Location = location;
```
Comment (ladipro) [inline src/Build/Construction/ProjectMetadataElement.cs:112]: nit: I wonder if there's a slightly more elegant way of setting location for newly created elements than adding a Location setter. If you passed the location to `ProjectRootElement.CreateElement` which in turn would pass it to a new overload of  `XmlDocumentWithLocation.CreateElement`, you could set its `_reader` field for the duration of the base call, similar to what `Load` does. Note that `_reader` is of type `IXmlLineInfo`, so not necessarily an XML reader. Or you could introduce a new field (`_currentLocation`?) , which would then even avoid boxing.

Overall, I would find it nicer to leave the `*WithLocation` classes concerned with setting location without leaking the setter like this. What do you think?

Comment (JanKrivanek) [reply src/Build/Construction/ProjectMetadataElement.cs:112]: Thanks for looking into this! Great suggestion!

I was as well trying to see if I can just extend the creation and avoid subsequent mutation.
The problem that stopped me and let me to current code is that the `XmlDocumentWithLocation` extends `XmlDocument` and our overriden `CreateElement` is not being called directy, but via a `XmlDocument` function that inside calls internal helper function:

image


So we'd need to duplicate some of that internal `XmlDocument` functionality (and maintain it going forward).

Not sure which is the lesser evil (allowing mutations vs duplicating `XmlDocument` code) - I arbitrarily decided for openng for mutation.

What do you think about this?

Comment (JanKrivanek) [reply src/Build/Construction/ProjectMetadataElement.cs]: Great thanks for offline discussion and tips! - I ended up passing the info via `AsyncLocal` filed.

**Review by ladipro — APPROVED** [review]
Looks great! I'm curious why you chose `AsyncLocal<>` over `ThreadLocal<>` when the relevant code is not async.

**Review by rainersigwald — APPROVED** [review]
Is there a way to add some test coverage for this?

**Comment by JanKrivanek** [comment]
> Is there a way to add some test coverage for this?

Tests added

**Pull request merged**